src/main/java/com/coreos/jetcd/Maintenance.java

```diff
@@ -64,36 +66,47 @@
   ListenableFuture<StatusResponse> statusMember();
 
   /**
-   * Set callback for snapshot.
+   * snapshot provides a reader for a snapshot of a backend.
    *
-   * <p>The onSnapshot will be called when the member make a snapshot.
+   * <pre>
+   * {code
+   * // create snapshot.db file current folder.
+   * String dir = Paths.get("").toAbsolutePath().toString();
+   * File snapfile = new File(dir, "snapshot.db");
    *
-   * <p>The onError will be called as exception, and the callback will be canceled.
+   * // leverage try-with-resources since SnapshotReader implements Closeable.
+   * try(SnapshotReader snapshotReader = maintenance.snapshot()){
+   *    while (true) {
+   *    SnapshotReaderResponse snapshotReaderResponse = snapshotReader.read();
+   *    if (snapshotReaderResponse.remaining == -1) {
+   *      break;
+   *    }
+   *    Files.write(snapfile.toPath(), snapshotReaderResponse.data);
+   *    }
+   * } catch (Exception e) {
+   * }
+   * }
+   * </pre>
    *
-   * @param callback Snapshot callback
+   * @return @see SnapshotReader
    */
-  void setSnapshotCallback(SnapshotCallback callback);
+  SnapshotReader snapshot();
 
-  /**
-   * Remove callback for snapshot.
-   */
-  void removeSnapShotCallback();
-
-  /**
-   * Callback to process snapshot events.
-   */
-  interface SnapshotCallback {
+  interface SnapshotReader extends Closeable {
 
     /**
-     * The onSnapshot will be called when the member make a snapshot.
+     * Returns a SnapshotReaderResponse.
      *
-     * @param snapshotResponse snapshot response
-     */
-    void onSnapShot(SnapshotResponse snapshotResponse);
-
-    /**
-     * The onError will be called as exception, and the callback will be canceled.
+     * <p>this method blocks until farther snapshot data are available,
+     * end of stream is detected, or an exception is thrown.
+     *
+     * <p>when there are no snapshot data to be read,
+     * a SnapshotReaderResponse with <code>remaining</code> -1 is returned.
+     *
+     * @return @see com.coreos.jetcd.maintenance.SnapshotReaderResponse
+     * @throws IOException if connection issue, SnapshotReader closed, any I/O issues.
      */
-    void onError(Throwable throwable);
+    SnapshotReaderResponse read() throws IOException;
```

Comment:
doesn't need to implement `InputStream` directly (looked too complicated, but maybe there's a shortcut?)

it could be something like:
```java
class Snapshot {
    ...
    public void write(OutputStream os) throws IOException;
    ...
};
```
can get an `InputStream` out of `write` by using `PipedInputStream`